### PR TITLE
log

### PR DESCRIPTION
**Pull request by zeitlinger**

### CODE REVIEW

File: engine/src/engine.ts [hunk 1]
```diff
@@ -105,41 +106,44 @@ export interface LogEntry {
 }
 
 const replaceRegex = new RegExp(
-  `\\b((${Command.Pass}|${Building.GaiaFormer}|${Command.UpgradeResearch}) ?([^. ]+)?)(\\.|$)`,
+  `\\b((${Command.Pass}|${Building.GaiaFormer}|${Command.FormFederation} [^ ]+|${Command.UpgradeResearch}) ?([^. ]+)?)(\\.|$)`,
```
Comment (coyotte508) [inline engine/src/engine.ts:109]: Like for the others - add a test to make sure that adding extra strings after `Command.FormFederation` doesn't mess up the replays

**Review by coyotte508 — CHANGES_REQUESTED** [review]
I really like the refactoring on the power tokens - and also how easy it is now to add power tokens info on federations

File: engine/src/player-data.ts [hunk 2]
```diff
@@ -450,7 +439,16 @@ export default class PlayerData extends EventEmitter {
     this.power.area1 -= area1ToGaia;
     this.power.area2 -= area2ToGaia;
     this.power.area3 -= area3ToGaia;
-    this.power.gaia += area1ToGaia + area2ToGaia + area3ToGaia;
+
+    const event: MoveTokens = {
+      area1: area1ToGaia,
+      area2: area2ToGaia,
+      area3: area3ToGaia,
+      gaia: 0,
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply engine/src/player-data.ts:447]: it actually removes them - renamed to `discardTokens`

File: engine/src/player-data.ts [hunk 3]
```diff
@@ -450,7 +439,16 @@ export default class PlayerData extends EventEmitter {
...
+      brainstone: movedBrainstone,
+    };
+    this.emit("move-tokens", event);
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply engine/src/player-data.ts:450]: it's a modifier functions

File: viewer/tsconfig.json [hunk 4]
```diff
@@ -2,7 +2,7 @@
   "compilerOptions": {
     /* Basic Options */
     "target": "es2019",
-    "module": "ESNext",
+    "module": "CommonJS",
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply viewer/tsconfig.json:5]: thanks :smile:

File: engine/src/player-data.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (coyotte508) [reply engine/src/player-data.ts:450]: My bad. I didn't notice that this was kept:

```
    this.power.area1 -= area1ToGaia;
    this.power.area2 -= area2ToGaia;
    this.power.area3 -= area3ToGaia;
```

File: engine/src/engine.ts [hunk 5]
```diff
@@ -118,9 +118,11 @@ export function createMoveToShow(move: string, p: PlayerData, executeMove: () =>
 
   const formerBooster = p.tiles.booster;
 
-  executeMove();
-
-  p.off("move-tokens", listener);
+  try {
```
Comment (coyotte508) [inline engine/src/engine.ts:121]: I'd have moved `p.on(...)` just above the `try` (I'm a bit paranoid)

Comment (zeitlinger) [reply engine/src/engine.ts:121]: I agree - it's a good pattern (will change later)

**Pull request merged**